### PR TITLE
[Doc] Ray logging document.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

An initial draft of the logging document.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-logging.rst [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    ray.init()
+
+    @ray.remote
+    class Actor:
+        def __init__(self):
+            # Basic config automatically configures logs to
+            # be streamed to stdout and stderr.
+            logging.basicConfig(level=logging.DEBUG)
+        
+        def log(self, msg):
```
Comment (edoakes) [inline doc/source/ray-logging.rst:61]: This doesn't address how to setup up logging for tasks (I'm not actually sure how to do this)

File: doc/source/ray-logging.rst [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    (pid=45601) task
+
+How to set up loggers
+~~~~~~~~~~~~~~~~~~~~~
+When using ray, all of the tasks and actors are executed remotely in Ray's worker processes. 
+Since Python logger module creates a singleton logger per process, loggers should be configured on per task/actor basis. 
+
+.. note::
+
+    Note that to stream logs to a driver, they should be flushed to stdout and stderr.
```
Comment (architkulkarni) [inline doc/source/ray-logging.rst:45]:
```suggestion
    To stream logs to a driver, they must be flushed to stdout and stderr.
```

Comment (architkulkarni) [reply doc/source/ray-logging.rst:45]: Just because it's already in a "Note"

File: doc/source/ray-logging.rst [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+    The default temp directory is `/tmp/ray` (for Linux and Mac OS). If you'd like to change the temp directory, you can specify it when `ray start` or `ray.init()` is called. 
+
+A new ray instance creates a new session ID to the temp directory. The latest session ID is symlinked to `/tmp/ray/session_latest`.
+
+Here's a Ray log directory structure. Note that `.out` is logs from stdout/stderr and `.err` is logs fro stderr. The backward compatibility of log directories is not maintained.
+
+- `dashboard.log`: A log file of a ray dashboard.
+- `dashboard_agent.log`: Every ray node has one dashboard agent. This is a log file of the agent.
+- `gcs_server.[out|err]`: GCS server is a stateless server that manages business logic that needs to be performed on GCS (Redis). It exists only in a head node.
```
Comment (architkulkarni) [inline doc/source/ray-logging.rst:108]:
```suggestion
- `gcs_server.[out|err]`: The GCS server is a stateless server that manages business logic that needs to be performed on GCS (Redis). It exists only in the head node.
```

File: doc/source/ray-logging.rst [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+- `log_monitor.log`: Log monitor is in charge of streaming logs to a driver.
```
Comment (architkulkarni) [inline doc/source/ray-logging.rst:109]:
```suggestion
- `log_monitor.log`: The log monitor is in charge of streaming logs to the driver.
```

File: doc/source/ray-logging.rst [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+- `monitor.log`: Ray's cluster launcher is operated with a monitor process. It also manages autoscaler.
```
Comment (architkulkarni) [inline doc/source/ray-logging.rst:110]:
```suggestion
- `monitor.log`: Ray's cluster launcher is operated with a monitor process. It also manages the autoscaler.
```

File: doc/source/ray-logging.rst [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+- `monitor.[out|err]`: Stdout and stderr of a cluster launcher.
+- `plasma_store.[out|err]`: Deprecated.
+- `python-core-driver-[worker_id]_[pid].log`: Ray drivers consist of CPP core and Python/Java frontend. It is a log file generated from CPP code.
```
Comment (architkulkarni) [inline doc/source/ray-logging.rst:113]:
```suggestion
- `python-core-driver-[worker_id]_[pid].log`: Ray drivers consist of CPP core and Python/Java frontend. This is a log file generated from CPP code.
```

File: doc/source/ray-logging.rst [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+- `python-core-worker-[worker_id]_[pid].log`: Ray workers consist of CPP core and Python/Java frontend. It is a log file generated from CPP code.
```
Comment (architkulkarni) [inline doc/source/ray-logging.rst:114]:
```suggestion
- `python-core-worker-[worker_id]_[pid].log`: Ray workers consist of CPP core and Python/Java frontend. This is a log file generated from CPP code.
```

File: doc/source/ray-logging.rst [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+- `raylet.[out|err]`: A log file of raylets.
+- `redis-shard_[shard_index].[out|err]`: A log file of GCS (Redis by default) shards.
+- `redis.[out|err]`: A log file of GCS (Redis by default).
+- `worker-[worker_id]-[job_id]-[pid].[out|err]`: Python/Java part of Ray drivers and workers. All of stdout and stderr from tasks/actors are streamed here. Note that job_id is an id of the driver.
+- `io-worker-[worker_id]-[pid].[out|err]`: Ray creates IO workers to spill/restore objects to external storage by default from Ray 1.3+. It is a log file of IO workers.
```
Comment (architkulkarni) [inline doc/source/ray-logging.rst:119]:
```suggestion
- `io-worker-[worker_id]-[pid].[out|err]`: Ray creates IO workers to spill/restore objects to external storage by default from Ray 1.3+. This is a log file of IO workers.
```

**Review by architkulkarni — CHANGES_REQUESTED** [review]
This looks great and very clear!  I just added some minor suggestions for the text.

File: doc/source/ray-logging.rst [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    import logging
+    # Initiate a driver.
+    ray.init()
+
+    @ray.remote
+    class Actor:
+        def __init__(self):
+            # Basic config automatically configures logs to
+            # be streamed to stdout and stderr.
+            logging.basicConfig(level=logging.DEBUG)
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply doc/source/ray-logging.rst:59]: Nope. But I think the default severity is WARNING, so if you don't set this, it won't print anything.

File: doc/source/ray-logging.rst (hunk 1 above)
Comment (rkooo567) [reply doc/source/ray-logging.rst:61]: Done.

**Review by rkooo567 — COMMENTED** [review]
@architkulkarni Thanks for the detailed review! :)

File: doc/source/ray-logging.rst [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+Driver logs
+~~~~~~~~~~~
+An entry point of Ray applications that calls `ray.init(address='auto')` or `ray.init()` is called a driver.
+All the driver logs are handled in the same way as normal python programs. 
+
+Worker logs
+~~~~~~~~~~~
+Ray's tasks or actors are executed remotely within Ray's worker processes. Ray has special support to improve the visibility of logs produced by workers.
+- By default, all of the tasks/actors stdout and stderr are redirected to the worker log files. Check out :ref:`Logging directory structure <logging-directory-structure>` to learn how Ray's logging directory is structured.
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply doc/source/ray-logging.rst:14]: I tried making it okay. Let's see how it looks.

**Review by architkulkarni — APPROVED** [review]

**Pull request merged**